src/config/mod.rs

```diff
@@ -0,0 +1,195 @@
+pub struct HaConfig {
+    pub host: String,
+    #[serde(rename = "long-lived-token")]
+    pub long_lived_token: Option<String>,
+    #[serde(rename = "device-id")]
+    pub device_id: Option<String>,
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone)]
+pub struct YamlConfig {
+    pub ha: HaConfig,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct WsAuthRequest {
+    #[serde(rename(serialize = "type"))]
+    auth_type: String,
+    access_token: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct WsLongLivedAccessTokenRequest {
+    id: u32,
+    #[serde(rename(serialize = "type"))]
+    command_type: String,
+    client_name: String,
+    lifespan: u32,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct WsLongLivedAccessTokenResponse {
+    id: u32,
+    #[serde(rename(deserialize = "type"))]
+    command_type: String,
+    success: bool,
+    result: Option<String>,
+}
+
+type Result<T> = std::result::Result<T, Box<dyn error::Error>>;
+
+pub async fn wait_for_token(config: &YamlConfig) -> Result<GetAccessTokenResponse> {
+    let server = Server::http("0.0.0.0:8000").unwrap();
+    let mut get_token_response: Option<GetAccessTokenResponse> = None;
+    println!("Open http://{}/auth/authorize?client_id=http%3A%2F%2Flocalhost%3A8000&redirect_uri=http%3A%2F%2Flocalhost%3A8000%2Fcallback in your browser", config.ha.host.as_str());
+    // blocks until the next request is received
+    match server.recv() {
+        Ok(request) => {
+            let url = format!("http://localhost:8000{}", request.url());
+            let query_params: HashMap<_, _> = Url::parse(url.as_str())
+                .unwrap()
+                .query_pairs()
+                .into_owned()
+                .collect();
+            match query_params.get("code") {
+                Some(code) => {
+                    request.respond(Response::from_string(
+                        "Halcyon now authenticated to Home Assistant. You can close this page now.",
+                    ))?;
+                    let either = ha_api::get_access_token(config, code.to_string()).await?;
+                    match either {
+                        either::Left(_) => {
+                            get_token_response = None;
+                        }
+                        either::Right(succ) => {
+                            get_token_response = Some(succ);
+                        }
+                    }
+                }
+                None => {
+                    request.respond(
+                        Response::from_string(
+                            "Something went wrong with authenticating Halcyon with Home Assistant",
+                        )
+                        .with_status_code(500),
+                    )?;
+                    get_token_response = None;
+                }
+            }
+        }
+        Err(e) => {
+            println!("error: {}", e);
+        }
+    };
+    get_token_response
+        .map(Ok)
```

Comment:
Is there a need to map this into a `Result`? Maybe have the caller deal with printing the error, and have the method just return `Option`